Improve Indices on pruned nodes via prune blockers

fjahr opened the pull request:
# Motivation
The main motivation of this change and only behavior change noticeable by user is to allow running `coinstatsindex` on pruned nodes as has been requested [here for example](https://twitter.com/benthecarman/status/1388170854140452870?s=20).

# Background
`coinstatsindex` on pruned nodes can be enabled in a much simpler than it is done here but it comes with downside. The ability to run `blockfilterindex`on pruned nodes was added in #15946 but it also added the `blockfilterindex` as a dependency to `validation` and it introduced two new circular dependencies. Enabling `coinstatsindex` on pruned nodes in a similar way would add it as a dependency as well and introduce another circular dependency. 

Instead, this PR introduces a `m_prune_blockers` map to `BlockManager` as a flexible approach to block pruning. Entities like `blockfilterindex`, for example, can add a key and a height to block pruning over that height. These entities need to update that value to allow more pruning when they are ready.

# Alternative approach
Upon completing the first draft of this PR I found #19463 as an alternative that follows the same but follows a very different approach. I am listing the main differences here as I see them:
- Usage of globals
- Blocks pruning with a start and a stop height
- Can persist blockers across restarts
- Blockers can be set/unset via RPCs

Personally, I don't think any of these are necessary to be added here but if the general approach or specific features are more appealing to reviewers I am happy to change to a solution based on that PR or port over specific parts of it here.

fjahr committed:
refactor: Introduce GetFirstStoredBlock helper function

fjahr committed:
Index: Use prune locks for blockfilterindex

Prior to this change blocks could be pruned up to the last block before the blockfilterindex current best block.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -1335,6 +1334,14 @@ int BlockManager::GetSpendHeight(const CCoinsViewCache& inputs)
 
+const CBlockIndex* BlockManager::GetLastPruneBlock(const CBlockIndex* start_block) {
+    const CBlockIndex* block = start_block;
+    CHECK_NONFATAL(block);
```
This function doesn't use any member of BlockManager, so could be stand-alone? Also, I think you can avoid this check by having the caller pass a `const CBlockIndex& start_block`.

fjahr replied:
Thanks! I made the changes as suggested.

fjahr committed:
blockstorage: Add prune locks to BlockManager

This change also introduces an aditional buffer of 10 blocks (PRUNE_LOCK_BUFFER) that will not be pruned before the best block.

Co-authored-by: Luke Dashjr <luke-jr+git@utopios.org>

fjahr committed:
Index: Allow coinstatsindex with pruning enabled

fjahr committed:
test: Update test for indices on pruned nodes

fjahr committed:
move-only: Rename index + pruning functional test

laanwj commented:
Concept ACK, I think this is an elegant approach.

achow101 commented on src/validation.cpp:
```diff
@@ -3663,6 +3673,10 @@ void BlockManager::FindFilesToPrune(std::set<int>& setFilesToPrune, uint64_t nPr
            nLastBlockWeCanPrune, count);
 }
 
+void BlockManager::UpdatePruneBlocker(std::string name, const CBlockIndex* block) {
```
In 5d23ddf5ff3bd1e5d4f6b85440d9ad0c59bb9994 "Add prune blockers to BlockManager"

Since this has `EXCLUSIVE_LOCKS_REQUIRED`, I think it should also have `AssertLockHeld` in here too.

fjahr replied:
Right, fixed!

achow101 reviewed (commented):
ACK 11f01efadf8e93f042ebd05d12f1f3bc5bce8178

promag commented on src/node/blockstorage.cpp:
```diff
@@ -51,6 +51,14 @@ bool IsBlockPruned(const CBlockIndex* pblockindex)
     return (fHavePruned && !(pblockindex->nStatus & BLOCK_HAVE_DATA) && pblockindex->nTx > 0);
 }
 
+const CBlockIndex* GetLastPruneBlock(const CBlockIndex& start_block) {
```
bfe9719e29d7947e2ffa9b510cd078e8a09b5d03

nit, `GetLastPrunedBlock`?

fjahr replied:
done

promag commented on src/validation.h:
```diff
@@ -450,6 +450,10 @@ class BlockManager
 
+    std::unordered_map<std::string, const CBlockIndex*> m_prune_blockers;
+
+    void UpdatePruneBlocker(std::string name, const CBlockIndex* block) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
```
b5a239390fd1b31c7e6107a2e6c35658a2b4697c

nit, `const std::string&`.

fjahr replied:
done

promag commented on src/validation.cpp:
```diff
@@ -2041,6 +2041,16 @@ bool CChainState::FlushStateToDisk(
             });
 
+            for (auto const& blocker : m_blockman.m_prune_blockers) {
+                if (blocker.second) {
```
b5a239390fd1b31c7e6107a2e6c35658a2b4697c

How can it be `nullptr`?

fjahr replied:
I guess it should never be `nullptr`. Dropped that if statement.

promag commented on src/validation.cpp:
```diff
@@ -3663,6 +3673,11 @@ void BlockManager::FindFilesToPrune(std::set<int>& setFilesToPrune, uint64_t nPr
 
+void BlockManager::UpdatePruneBlocker(std::string name, const CBlockIndex* block) {
+    AssertLockHeld(::cs_main);
+    m_prune_blockers[name] = block;
```
b5a239390fd1b31c7e6107a2e6c35658a2b4697c

Should check/assert it only goes forward when updating?

fjahr replied:
I am not sure. The way I use it right now it can also go down with reorgs. I am undecided what the right course is. Only going up avoids a potential footgun, with going down it could react to (highly unlikely) deep reorg. If I add this I think I would do it so that setting a lower number is simply a no op:

```
if (m_prune_blockers[name] < block) m_prune_blockers[name] = block;
```

promag replied:
But what should happen if the block is already pruned?

fjahr replied:
Hm, that would mean the node would have to restart with `-reindex` anyway, right?

fjahr replied:
Yes, I added functional tests for the -reindex case.

I am looking into the sanity check assert idea don't have a clear opinion yet.

promag commented on src/index/base.cpp:
```diff
@@ -364,3 +364,15 @@ IndexSummary BaseIndex::GetSummary() const
+    if (AllowPrune()) {
+        LOCK(::cs_main);
+        if (!block) {
+            g_chainman.m_blockman.UpdatePruneBlocker(GetName(), ::ChainActive().Genesis());
```
d18677006d3d48ca7654039ea6e267164a2690e0

nit, use ternary? `block ? block : ::ChainActive().Genesis()`

fjahr replied:
done

promag reviewed (commented):
Concept ACK.

fjahr commented:
Addressed comments by @promag , thank you!

I also fixed a comment I had overlooked and added a new commit at the end because I realized that we can skip the pruning stuff in `index/base` if the node is not actually pruning. With the new behavior especially this prevents locking `cs_main`. Happy to squash this but I am still checking that this is 100% safe and want to make sure it is looked at separately because it was not included in the existing concept acks.

promag reviewed (commented):
How about keeping some blocks before the deepest prune blocker?

fjahr commented:
> How about keeping some blocks before the deepest prune blocker?

Interesting. That should help make it a bit more robust. I added this but set it pretty low for now (10), looking for feedback if it should be higher. I also encapsulated the logic for getting the deepest prune height into it's own function to not have more logic in `FlushStateToDisk`.

Also rebased.

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -1069,12 +1069,9 @@ static RPCHelpMan pruneblockchain()
-        block = block->pprev;
-    }
-    return uint64_t(block->nHeight);
+    const CBlockIndex* last_block = GetLastPrunedBlock(*active_chain.Tip());
```
In commit "refactor: Introduce GetLastPruneBlock helper function" (292d508e956bfe5b05e6ee685415063c2a0f04af)

Note: Here and in the other location below there's no more CHECK_NONFATAL so there would be segfault instead of an exception if the tip was null. Doesn't seem like a problem, but if were for some reason, GetLastPrunedBlock argument could be a pointer instead of a reference and the check could be added there.

fjahr replied:
True, I changed the function to take a pointer and added a CHECK_NONFATAL to be more consistent with the previous code.

ryanofsky reviewed (commented):
Approach ACK and started review (will update list below with progress).

- [X] 292d508e956bfe5b05e6ee685415063c2a0f04af refactor: Introduce GetLastPruneBlock helper function (1/7)
- [x] 9f8d6a9a8f7d113ae12306c72a773830b5725e37 Add prune blockers to BlockManager (2/7)
- [ ] c07a567835ca3232fa30d4e5306592e23d55dacd Index: Use prune blockers for blockfilterindex (3/7)
- [ ] 395757f3080b96baec65e42cb5e56051a17f2241 Index: Allow coinstatsindex with pruning enabled (4/7)
- [ ] 6aad8676952f2a854e9f8bf8650bb7768d3e1cd1 test: Update test for indices on pruned nodes (5/7)
- [ ] 596f738ec33efb40e6e2eee27356e493e85bf8ed move-only: Rename index + pruning functional test (6/7)
- [ ] e365e87a72cc49e175b31256a11a6c5b1621d1f2 Index: Skip pruning checks when node is not pruning (7/7)

ryanofsky commented on src/validation.cpp:
```diff
@@ -3792,6 +3794,21 @@ void BlockManager::FindFilesToPrune(std::set<int>& setFilesToPrune, uint64_t nPr
+        const int blocker_height = blocker.second->nHeight + PRUNE_BLOCKER_BUFFER;
+        last_prune = std::max(1, std::min(last_prune, blocker_height));
+        if (last_prune == blocker_height) {
+            LogPrintf("%s limited pruning to height %d\n", blocker.first, blocker_height);
```
In commit "Add prune blockers to BlockManager" (9f8d6a9a8f7d113ae12306c72a773830b5725e37):

Would it make sense to use a log category here `LogPrint(BCLog::PRUNE, ...)` instead of printing unconditionally? Also in the future it could be nice to unify this print with the prints in the FindFilesToPrune functions and print the full pruning status in a consistent way on just one line.

fjahr replied:
Yeah, I agree, I added a log category. I am leaving unifying for a potential follow-up to not complicate this PR further.

dergoegge commented on test/functional/feature_index_prune.py:
```diff
@@ -0,0 +1,117 @@
+
+        # Mine two batches of blocks to avoid hitting NODE_NETWORK_LIMITED_MIN_BLOCKS disconnection
+        self.mine_batches(2)
+        self.sync_index(height=700)
```
Wouldn't it make more sense to wait for the indexes to sync *after* pruning?
Otherwise it is pretty much guaranteed that the index data at the tip (700) is available after pruning.

In general i think it would be good for the test to check for expected prune blocker heights somehow. (not sure if that is possible, i dont know that much about the functional tests)

fjahr replied:
> Wouldn't it make more sense to wait for the indexes to sync _after_ pruning?
> Otherwise it is pretty much guaranteed that the index data at the tip (700) is available after pruning.

Yes, that would be good, but unless I am misunderstanding your comment, what we would be testing then is a race condition which would make the test flaky.

> In general i think it would be good for the test to check for expected prune blocker heights somehow. (not sure if that is possible, i dont know that much about the functional tests)

yes, I added two sanity checks for the exact prune blocker height by checking the logs

dergoegge reviewed (commented):
Concept ACK

jonatack commented on src/index/base.cpp:
```diff
@@ -370,3 +366,11 @@ IndexSummary BaseIndex::GetSummary() const
+    m_best_block_index = block;
+    if (AllowPrune() && (fPruneMode || fHavePruned) && !fReindex) {
+        LOCK(::cs_main);
+        g_chainman.m_blockman.UpdatePruneBlocker(GetName(), block ? block : ::ChainActive().Genesis());
```
Error building this branch at e365e87a72cc4 rebased to current master at 8cdf91735f2b, looks like it needs updating:
```rake
index/base.cpp:373:9: error: use of undeclared identifier 'g_chainman'
        g_chainman.m_blockman.UpdatePruneBlocker(GetName(), block ? block : ::ChainActive().Genesis());
        ^
index/base.cpp:373:79: error: no member named 'ChainActive' in the global namespace
        g_chainman.m_blockman.UpdatePruneBlocker(GetName(), block ? block : ::ChainActive().Genesis());
                                                                            ~~^
2 errors generated.
```

fjahr replied:
Thanks, fixed.

jonatack commented:
Nudge to update.

MarcoFalke commented:
Removed from high-prio for now due to inactivity, but can be added back

fjahr commented on src/validation.cpp:
```diff
@@ -3792,6 +3790,21 @@ void BlockManager::FindFilesToPrune(std::set<int>& setFilesToPrune, uint64_t nPr
+
+void BlockManager::GetLastPruneBlockerHeight(int& last_prune) {
+    for (auto const& blocker : m_prune_blockers) {
+        const int blocker_height = blocker.second->nHeight + PRUNE_BLOCKER_BUFFER;
```
right 🙄 thanks, fixed!

fjahr commented on src/validation.h:
```diff
@@ -494,6 +497,12 @@ class BlockManager
      */
     int GetSpendHeight(const CCoinsViewCache& inputs) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
+    std::unordered_map<std::string, const CBlockIndex*> m_prune_blockers;
```
Added!

fjahr commented on src/validation.cpp:
```diff
@@ -3792,6 +3794,21 @@ void BlockManager::FindFilesToPrune(std::set<int>& setFilesToPrune, uint64_t nPr
+    m_prune_blockers[name] = block;
+}
+
+void BlockManager::GetLastPruneBlockerHeight(int& last_prune) {
```
Done

fjahr commented:
Sorry for the long wait, I should have addressed all the feedback for now and also optimized the test a bit further.

ryanofsky commented:
> Sorry for the long wait, I should have addressed all the feedback for now and also optimized the test a bit further.

Thanks for updating this! I want to review it more soon, and it would be great to have feedback also from @jamesob because I think there is or could be some interaction between this and https://github.com/bitcoin/bitcoin/pull/15606#pullrequestreview-692965905

jamesob commented:
> it would be great to have feedback also from @jamesob

Yep, will plan to review this in the next few days.

jamesob reviewed (approved):
ACK 983a3552f22678d4466e0faa8e784f25b145a76b ([`jamesob/ackr/21726.1.fjahr.improve_indices_on_prune`](https://github.com/jamesob/bitcoin/tree/ackr/21726.1.fjahr.improve_indices_on_prune))

This change has two immediate benefits: (i) it removes an existing cicular depdendency between indexing code and validation, and (ii) allows use of coinstatsindex alongside pruning configuration. This inteface is also a nice way to decouple pruning itself from various consumers of blockdata that limit it.

<details><summary>Show signature data</summary>
<p>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 983a3552f22678d4466e0faa8e784f25b145a76b ([`jamesob/ackr/21726.1.fjahr.improve_indices_on_prune`](https://github.com/jamesob/bitcoin/tree/ackr/21726.1.fjahr.improve_indices_on_prune))

This change has two immediate benefits: (i) it removes an existing cicular depdendency between indexing code and validation, and (ii) allows use of coinstatsindex alongside pruning configuration. This inteface is also a nice way to decouple pruning itself from various consumers of blockdata that limit it.
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEGNRVI1NPYuZCSIrGepNdrbLETwUFAmE6KhgACgkQepNdrbLE
TwWm+A//e7YHR1ZGAp+nsrIF2ZtO0a/TSlR1YEVIwzcYXXmA6BonfRfS8OCXuKfa
cdKsbxQudd8vjPoQ+ec4jCKqep+1DxrZzes56Z1sinMYqk8W1IosMJYXFH+QZt4h
8ujmZ5bvtjBBStmpWW7jkUljGQlE2dMhH3h+jZ5F7P5nFDBgMeZoutUntgTkSWpk
G8wv5OMGgm5jNFXBfgyA7wi0SCMw/QNBhu/dKDbcCYg9CVDPh6sPXeuYAJPk2Xbd
3W5sLv+9cRFRAeQLeBc5112UnixqF4VCUJZIeLTcwgb1F786kAWj9SBurM0o0KKW
5AjXL2dmbmni3qxqLh3xAPjIXLJse+ZIVZ3S4LW8ds36jpY2oDtt4hzQb1Lf/bm3
bfvjuMz6SF25oZWnE+XUeg/kREb0JK4wxlr7hK3qIjvp8LxEf9vSmozEsIbNZZlB
q4o/+aFUuRy0mJPrJb7D5RqdC+/LIRXbkPSyeJiibYxnhOMAUK8d2i4atpfjpzN6
PF+8Es7GaUYbZnJxO6hxbRrEWimewnvRHS9N4jwDcFB0FRuFVWJmRfkcBwY8NTm3
5O22Abo7sPAWztZjOlM6yq6Vn1yuwVvFPSCzpE/eWZtAhhf86b7Dwnnd/NXiKa5s
ZzuhCuHQLGZJnX3sM38h/E9g4nmClP0Drf0uxDJNqyXw8nQIonQ=
=UW7v
-----END PGP SIGNATURE-----

```

</p></details>

<details><summary>Show platform data</summary>
<p>

```
Tested on Linux-4.19.0-17-amd64-x86_64-with-glibc2.28

Configured with ./configure LDFLAGS=-L/home/james/src/bitcoin/db4/lib/ CPPFLAGS=-I/home/james/src/bitcoin/db4/include/ CXXFLAGS=-fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare --enable-wallet --with-daemon CXX=/usr/local/bin/clang++ CC=/usr/local/bin/clang PYTHONPATH= --disable-shared --with-pic --enable-benchmark=no --with-bignum=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental --no-create --no-recursion

Compiled with /usr/bin/ccache /usr/local/bin/clang++ -std=c++17 -mavx -mavx2 -fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -fdebug-prefix-map=$(abs_srcdir)=.  -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -msse4 -msha -msse4.1 -msse4.2  i

Compiler version: Debian clang version 11.1.0-++20210428103820+1fdec59bffc1-1~exp1~20210428204437.162
```

</p></details>

dergoegge commented on test/functional/feature_index_prune.py:
```diff
@@ -0,0 +1,130 @@
+        self.mine_batches(2)
+        self.sync_index(height=700)
+
+        self.expected_prune_height(690)
```
I am able to put any value here besides 690 and the test still passes.

This is because [`assert_debug_log`](https://github.com/bitcoin/bitcoin/blob/5c0f46ca46e23a161649b5150baa01020dc85e48/test/functional/test_framework/test_node.py#L381) is a generator and meant to be used in combination with `with`:
```python
with assert_debug_log(...):
  # some other code that prints the (un)expected logs
```
"At the point where the generator yields, the block nested in the [`with`](https://docs.python.org/3/reference/compound_stmts.html#with) statement is executed. The generator is then resumed after the block is exited. [...]" from https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager

Secondly i think that checking the logs to test the prune blocker height won't work here as the expected message is only printed when pruning is attempted (https://github.com/bitcoin/bitcoin/blob/983a3552f22678d4466e0faa8e784f25b145a76b/src/validation.cpp#L2027)

the following would work:
```python
 self.log.info("prune some blocks")
 for node in self.nodes[:2]:
     with node.assert_debug_log(['limited pruning to height 690']):
         pruneheight = node.pruneblockchain(400)
         assert_equal(pruneheight, 248)
```

fjahr replied:
Thanks a lot for the detailed feedback! I fixed this using your suggested approach in the two locations where I wanted to do the check.

dergoegge reviewed (commented):
It's a bit sad that there isn't a way to pause/resume an index for testing purposes. With that the test here could do something like:
1. mine some blocks and let the index sync to tip
2. pause the index
3. mine a bunch of blocks
4. prune and check that the last pruned block (returned from `pruneblockchain`) matches the prune blocker.
5. resume the index and sync to tip
6. prune again and check that blocks after the previous blocker were pruned

might be something for a follow-up.

ryanofsky commented on src/node/blockstorage.h:
```diff
@@ -48,6 +48,9 @@ extern uint64_t nPruneTarget;
 //! Check whether the block associated with this index entry is pruned or not.
 bool IsBlockPruned(const CBlockIndex* pblockindex);
 
+//! Find the last block that is pruned
+const CBlockIndex* GetLastPrunedBlock(const CBlockIndex* start_block);
```
In commit "refactor: Introduce GetLastPruneBlock helper function" (2c815adc881fa90f14887e2ec33ccb11ef964387)

This doesn't seem to actually return the last block that is pruned. It seems to return the next block that has data. So maybe should be called `GetFirstNonPrunedBlock`? It also seems like `pruneblockchain` documentation describes this incorrectly while `getblockchaininfo` describes it correctly.

fjahr replied:
True, I fixed this and went with with the name `GetFirstStoredBlock` inspired by the docs in `getblockchaininfo`. I also added an extra commit to fix the docs in `pruneblockchain`. I think fixing the docs is the right approach rather than change the behavior but I kept it in the separate commit in case there are different opinions on that.

fjahr commented on src/validation.h:
```diff
@@ -488,6 +490,16 @@ class BlockManager
+     * @note Internally, only blocks at height (block->nHeight - PRUNE_BLOCKER_BUFFER) and
+     * below will be pruned, but callers should avoid assuming any particular buffer size.
+     */
+    std::unordered_map<std::string, const CBlockIndex*> m_prune_blockers;
```
Not intentional, just an oversight. Fixed!

fjahr commented on test/functional/feature_index_prune.py:
```diff
@@ -0,0 +1,130 @@
+        self.log.info("prune further than the indices best blocks while the indices are disabled")
+        for i in range(3):
+            pruneheight_new = self.nodes[i].pruneblockchain(1000)
+            assert_greater_than(pruneheight_new, pruneheight)
```
Makes sense, fixed!

ryanofsky commented on src/validation.cpp:
```diff
@@ -1948,6 +1948,15 @@ bool CChainState::FlushStateToDisk(
             });
 
+            for (auto const& blocker : m_blockman.m_prune_blockers) {
+                const int blocker_height = blocker.second->nHeight - PRUNE_BLOCKER_BUFFER;
```
In commit "Index: Use prune blockers for blockfilterindex" (7db7f26e97665c048b3f614c092899addd658f4f)

I'm seeing a segfault this line starting in the "Index: Use prune blockers" commit and continuing up to the ""Index: Skip pruning checks" commit in the `feature_blockfilterindex_prune.py` test because `blocker.second` is null. It would probably be good to not insert null prune blockers or just ignore them.

```
#0  0x000055555583d8a6 in CChainState::FlushStateToDisk (this=0x555555e46f10, state=..., mode=FlushStateMode::IF_NEEDED, nManualPruneHeight=0) at validation.cpp:1948
#1  0x0000555555849e51 in CChainState::ConnectTip (this=0x555555e46f10, state=..., pindexNew=0x7fffac001000, pblock=..., connectTrace=..., disconnectpool=...) at validation.cpp:2288
#2  0x000055555584f7f2 in CChainState::ActivateBestChainStep (this=0x555555e46f10, state=..., pindexMostWork=0x7fffac001000, pblock=std::shared_ptr<const CBlock> (empty) = {...}, fInvalidFound=@0x7fffa7ffe32f: false, connectTrace=...)
    at validation.cpp:2432
#3  0x000055555584fff8 in CChainState::ActivateBestChain (this=<optimized out>, state=..., pblock=std::shared_ptr<const CBlock> (empty) = {...}) at validation.cpp:2557
#4  0x00005555556c77fd in ThreadImport (chainman=..., vImportFiles=std::vector of length 0, capacity 0, args=...) at node/blockstorage.cpp:555
#5  0x0000555555633850 in operator() (__closure=0x7fffac000b60) at init.cpp:1656
#6  std::__invoke_impl<void, AppInitMain(NodeContext&, interfaces::BlockAndHeaderTipInfo*)::<lambda()>&> (__f=...) at /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/invoke.h:60
#7  std::__invoke_r<void, AppInitMain(NodeContext&, interfaces::BlockAndHeaderTipInfo*)::<lambda()>&> (__fn=...) at /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/invoke.h:110
#8  std::_Function_handler<void(), AppInitMain(NodeContext&, interfaces::BlockAndHeaderTipInfo*)::<lambda()> >::_M_invoke(const std::_Any_data &) (__functor=...)
    at /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/std_function.h:291
#9  0x0000555555b0ef5f in std::function<void ()>::operator()() const (this=0x7fffa7ffeb70) at /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/std_function.h:622
#10 util::TraceThread(char const*, std::function<void ()>) (thread_name=<optimized out>, thread_func=...) at util/thread.cpp:18
#11 0x000055555563376b in std::__invoke_impl<void, void (*)(char const*, std::function<void()>), char const*, AppInitMain(NodeContext&, interfaces::BlockAndHeaderTipInfo*)::<lambda()> > (
    __f=@0x555555e7f818: 0x555555b0edf0 <util::TraceThread(char const*, std::function<void ()>)>) at /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/invoke.h:60
#12 std::__invoke<void (*)(char const*, std::function<void()>), char const*, AppInitMain(NodeContext&, interfaces::BlockAndHeaderTipInfo*)::<lambda()> > (__fn=@0x555555e7f818: 0x555555b0edf0 <util::TraceThread(char const*, std::function<void ()>)>)
    at /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/invoke.h:95
#13 std::thread::_Invoker<std::tuple<void (*)(char const*, std::function<void()>), char const*, AppInitMain(NodeContext&, interfaces::BlockAndHeaderTipInfo*)::<lambda()> > >::_M_invoke<0, 1, 2> (this=0x555555e7f7e8)
    at /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/thread:264
#14 std::thread::_Invoker<std::tuple<void (*)(char const*, std::function<void()>), char const*, AppInitMain(NodeContext&, interfaces::BlockAndHeaderTipInfo*)::<lambda()> > >::operator() (this=0x555555e7f7e8)
    at /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/thread:271
#15 std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(char const*, std::function<void()>), char const*, AppInitMain(NodeContext&, interfaces::BlockAndHeaderTipInfo*)::<lambda()> > > >::_M_run(void) (this=0x555555e7f7e0)
    at /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/thread:215
#16 0x00007ffff7eaaca0 in ?? () from /nix/store/dzaadhn5y56da24635icafv1nrawxm4n-gcc-10.3.0-lib/lib/../lib64/libstdc++.so.6
#17 0x00007ffff7fb1e9e in start_thread () from /nix/store/gk42f59363p82rg2wv2mfy71jn5w4q4c-glibc-2.32-48/lib/libpthread.so.0
#18 0x00007ffff774949f in clone () from /nix/store/gk42f59363p82rg2wv2mfy71jn5w4q4c-glibc-2.32-48/lib/libc.so.6
```

fjahr replied:
Oh, TIL that `Genesis()` can return a `nullptr` and does so initially when reindexing. I never really checked and assumed this couldn't happen since the Genesis block is hardcoded. Well, good to know and this should be fixed now.

ryanofsky reviewed (commented):
Code review almost-ACK 70b9e4f6abda5d6dd38504d9149ddc79b33961ce. I reviewed most of this pretty thoroughly but I didn't really look at the python test yet. I think it would be good to fix the segfault in intermedate commits by not inserting null blockers, and to drop the "skip pruning checks" commit.

fjahr commented on src/index/base.cpp:
```diff
@@ -369,3 +365,12 @@ IndexSummary BaseIndex::GetSummary() const
     return summary;
 }
+
+void BaseIndex::SetBestBlockIndex(const CBlockIndex* block) {
```
Added

fjahr commented on src/validation.cpp:
```diff
@@ -1941,12 +1940,9 @@ bool CChainState::FlushStateToDisk(
             // pruning is height-based
             int last_prune = m_chain.Height(); // last height we can prune
-            ForEachBlockFilterIndex([&](BlockFilterIndex& index) {
-               last_prune = std::max(1, std::min(last_prune, index.GetSummary().best_block_height));
```
Added a note in the commit message

fjahr commented on src/index/base.cpp:
```diff
@@ -67,7 +67,7 @@ bool BaseIndex::Init()
     }
     m_synced = m_best_block_index.load() == active_chain.Tip();
-    if (!m_synced) {
+    if (!m_synced && (fPruneMode || fHavePruned) && !fReindex) {
```
I did not do performance tests on this change, it will probably not be huge but if we can get it, why not? But my main motivation is that I think that these checks are not necessary unless the conditions hold. I think it makes code easier to understand if sections that are not relevant are skipped explicitly. But I can see how this might be missing a comment for clarity, happy to add that if you agree.

fjahr replied:
I agree with you that the line is harder to understand after the change but in exchange for that, it makes explicit that the code inside the block is only relevant and needs to run if this evaluates to true. Let's wait for more feedback as a tie-breaker, of course, I am happy to remove it if others agree with you :)

dergoegge replied:
> it makes explicit that the code inside the block is only relevant and needs to run if this evaluates to true

I don't have a strong opinion but i agree with this statement and think the change should stay.

jonatack replied:
ISTM there should be a demonstrated clear motivation for the commit and documentation of that in both places.

fjahr commented:
Rebased and addressed @ryanofsky 's comments.

ryanofsky reviewed (approved):
Code review ~ACK f6766d436b22ac595f6e13d365d07af37dbc0ebd ignoring windows test failure. Changes since last review: rebasing and fixing GetIntArg conflicts, adding assert, fixing segfault in intermediate commits, and adding a description in one of the commit messages.

Windows test failure https://cirrus-ci.com/task/6252130924232704?logs=functional_tests#L2528 does seem like it is related. feature_index_prune.py is failing with "AssertionError: Block sync timed out after 480s." Unclear if that is just a timeout or if it is a real sync error.

fjahr commented:
Rebased and beefed up the wait time one last block sync that caused a CI failure previously. I think it was a normal timeout because I could not reproduce it locally and there are a lot of blocks being produced. Comparable tests in terms of number of blocks also use longer sync times.

I kindly ask other reviewers to weigh in on @ryanofsky 's feedback here if you have a preference: https://github.com/bitcoin/bitcoin/pull/21726#discussion_r715059159

jonatack commented:
Concept ACK, will have a look.

jonatack reviewed (commented):
Approach ACK 10eb269e04c3306ba3749e2d178253747319a565 with some suggestions. Rebased to latest master, reviewed, debug built and ran unit tests and relevant functional tests at each commit.

I had to run `test/functional/feature_index_prune.py --timeout-factor=0`, otherwise that updated test always times out for me locally even in turbo mode with the cpu maxed out.

<details><summary>test output</summary><p>

```
$ test/functional/feature_index_prune.py 
2021-10-06T15:47:31.133000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_289uzlfx
2021-10-06T15:47:32.445000Z TestFramework (INFO): check if we can access blockfilters and coinstats when pruning is enabled but no blocks are actually pruned
2021-10-06T15:48:32.497000Z TestFramework.utils (ERROR): wait_until() failed. Predicate: ''''
        self.wait_until(lambda: self.nodes[1].getindexinfo() == expected_stats)
'''
2021-10-06T15:48:32.498000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 131, in main
    self.run_test()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/feature_index_prune.py", line 54, in run_test
    self.sync_index(height=200)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/feature_index_prune.py", line 34, in sync_index
    self.wait_until(lambda: self.nodes[1].getindexinfo() == expected_stats)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 689, in wait_until
    return wait_until_helper(test_function, timeout=timeout, timeout_factor=self.options.timeout_factor)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/util.py", line 257, in wait_until_helper
    raise AssertionError("Predicate {} not true after {} seconds".format(predicate_source, timeout))
AssertionError: Predicate ''''
        self.wait_until(lambda: self.nodes[1].getindexinfo() == expected_stats)
''' not true after 60.0 seconds
2021-10-06T15:48:32.501000Z TestFramework (INFO): Stopping nodes
```
</p></details>

ryanofsky reviewed (approved):
Code review ACK 86121153164a05a5da9812010f0bfdab434bec1f. Only changes are fixing AddArg rebase conflict and adding sync_blocks python test timeout

MarcoFalke commented on test/functional/feature_index_prune.py:
```diff
@@ -0,0 +1,132 @@
+
+    def mine_batches(self, n):
+        for _ in range(n):
+            self.nodes[0].generate(250)
```
still not fixed?

fjahr replied:
Done

MarcoFalke commented:
Test failue: https://github.com/bitcoin/bitcoin/pull/21726#discussion_r745556769

fjahr commented on test/functional/test_runner.py:
```diff
@@ -310,7 +310,7 @@
     'feature_shutdown.py',
     'p2p_ibd_txrelay.py',
-    'feature_blockfilterindex_prune.py'
+    'feature_index_prune.py'
```
Hmm, 16 minutes? There must be something off. The test takes ~40 seconds on my (5+ year old) laptop. Can you check again if there wasn't another cause and maybe share some more info?

```
$ test/functional/feature_index_prune.py
2021-12-04T23:40:35.661000Z TestFramework (INFO): Initializing test directory /var/folders/9z/n7rz_6cj3bq__11k5kcrsvvm0000gn/T/bitcoin_func_test_t25asmxj
2021-12-04T23:40:36.639000Z TestFramework (INFO): check if we can access blockfilters and coinstats when pruning is enabled but no blocks are actually pruned
2021-12-04T23:40:43.562000Z TestFramework (INFO): prune some blocks
2021-12-04T23:40:43.566000Z TestFramework (INFO): check if we can access the tips blockfilter and coinstats when we have pruned some blocks
2021-12-04T23:40:43.595000Z TestFramework (INFO): check if we can access the blockfilter and coinstats of a pruned block
2021-12-04T23:40:43.615000Z TestFramework (INFO): restart nodes without indices
2021-12-04T23:40:44.986000Z TestFramework (INFO): make sure trying to access the indices throws errors
2021-12-04T23:40:53.843000Z TestFramework (INFO): prune further than the indices best blocks while the indices are disabled
2021-12-04T23:40:54.330000Z TestFramework (INFO): make sure we get an init error when starting the nodes again with the indices
2021-12-04T23:40:55.357000Z TestFramework (INFO): make sure the nodes start again with the indices and an additional -reindex arg
2021-12-04T23:41:14.996000Z TestFramework (INFO): Stopping nodes
2021-12-04T23:41:15.223000Z TestFramework (INFO): Cleaning up /var/folders/9z/n7rz_6cj3bq__11k5kcrsvvm0000gn/T/bitcoin_func_test_t25asmxj on exit
2021-12-04T23:41:15.223000Z TestFramework (INFO): Tests successful
```

jonatack replied:
Re-ran `time test/functional/feature_index_prune.py --timeout-factor=0` a few times, best times are 12 to 13 minutes with CPUs at the max setting (3100 MHz for 2500 MHz core x 2), and ~ a half hour at more leisurely CPU settings.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -1031,7 +1031,7 @@ static RPCHelpMan pruneblockchain()
                 },
                 RPCResult{
-                    RPCResult::Type::NUM, "", "Height of the last block pruned"},
+                    RPCResult::Type::NUM, "", "Height of the first block that is still stored after pruning"},
```
Done

fjahr commented on src/validation.h:
```diff
@@ -480,6 +482,16 @@ class BlockManager
+     * @note Internally, only blocks at height (block->nHeight - PRUNE_BLOCKER_BUFFER) and
+     * below will be pruned, but callers should avoid assuming any particular buffer size.
+     */
+    std::unordered_map<std::string, const CBlockIndex*> m_prune_blockers GUARDED_BY(::cs_main);
```
Done

fjahr commented on src/validation.h:
```diff
@@ -88,6 +88,8 @@ static const unsigned int DEFAULT_CHECKLEVEL = 3;
 // one 128MB block file + added 15% undo data = 147MB greater for a total of 545MB
 // Setting the target to >= 550 MiB will make it likely we can respect the target.
 static const uint64_t MIN_DISK_SPACE_FOR_BLOCK_FILES = 550 * 1024 * 1024;
+/** Blocks to keep below deepest prune blocker */
```
Done

fjahr commented on src/validation.cpp:
```diff
@@ -3594,6 +3599,11 @@ void BlockManager::FindFilesToPrune(std::set<int>& setFilesToPrune, uint64_t nPr
 
+void BlockManager::UpdatePruneBlocker(const std::string& name, const CBlockIndex* block) {
+    AssertLockHeld(::cs_main);
+    m_prune_blockers[name] = block;
```
Hm, I would like to keep this setter, it feels more clean to me.

fjahr commented:
Rebased and addressed all of @jonatack 's comments (I still can't figure out why GitHub lets me comment on and close some comments and not others).

I also decided to drop the last commit that @ryanofsky [wasn't very happy with](https://github.com/bitcoin/bitcoin/pull/21726#discussion_r719590304) in order to make the rest of the review process a bit easier hopefully. In general, I think the whole topic "are we pruning? do we have to do something about that or not?" is a bit messy and I will tackle that more holistically in a future PR.

dergoegge commented:
ACK deaa5f537bf21c33c8eb8ca0c276172a6a225f67

fjahr commented:
Another two rebases later...

Unfortunately there were some significant changes merged to the functional test, so there is a bit more to re-review this time 😢  Also these changes have more than doubled the runtime of the test for me. I have tried some simple optimizations but wasn't successful it getting a meaningful speedup so far. I will revisit speeding up the tests in a follow-up because I don't want to block review here further.

That is, unless the test is now too slow to run on some machines now. Please test it again on your toaster, @jonatack ;)

luke-jr commented on src/validation.cpp:
```diff
@@ -2222,12 +2222,18 @@ bool CChainState::FlushStateToDisk(
-            });
+
+            for (const auto& blocker : m_blockman.m_prune_blockers) {
+                const int blocker_height{blocker.second->nHeight - PRUNE_BLOCKER_BUFFER};
```
Can be fixed with something like 55f04cf1dd0f36981e3d09c180da52e11c974575

fjahr replied:
Conceptually i makes sense that the blockers can move backwards in reorg scenarios, so I have added this, though to be honest I have a hard time thinking of a realistic scenario where this prevents an issue on mainnet. I have also add a test for this feature.

luke-jr commented on src/validation.h:
```diff
@@ -455,6 +458,16 @@ class BlockManager
+     * @note Internally, only blocks at height (block->nHeight - PRUNE_BLOCKER_BUFFER) and
+     * below will be pruned, but callers should avoid assuming any particular buffer size.
+     */
+    std::unordered_map<std::string, const CBlockIndex*> m_prune_blockers GUARDED_BY(::cs_main);
```
Better to just use a height here, I think. See 598b0d8dbf5793c60c63dc2a181a49291fe8011b

fjahr replied:
Added this

luke-jr reviewed (changes requested):
Looked into using this to rebase #19463. Seems mostly fine, with one bug and a few nits.

The RPC interface for prune locks (not in this PR presently) could be useful for testing, so it might be a good idea to add it. If you want, you can just cherry-pick 9c9e0c0b8c6a5a332f45896261aca5bdb3746447 + e83ed21172049ef593702c52bb840868728f47d1 + da86f7bdcfda0f5b27091b3934bc9f3e8de03157 for that.

dongcarl commented:
Concept ACK, was just working on my own version of this when luke-jr alerted me to this PR. Thanks for doing the work so I can just cherry-pick haha :-)

One thing I'm wondering: have you considered having `CChainState` "pull" the `blocker_height` from blockers when necessary instead of requiring blockers to update `CChainState` every time their `blocker_height` changes?

luke-jr commented:
>One thing I'm wondering: have you considered having CChainState "pull" the blocker_height from blockers when necessary instead of requiring blockers to update CChainState every time their blocker_height changes?

That won't work for external or offline blockers. In #19463, my primary goal is to support tools like Armory or Lightning nodes that require parsing the block data, as well as ensuring pruning doesn't invalidate wallet backups (even of offline wallets).

dongcarl commented:
> > One thing I'm wondering: have you considered having CChainState "pull" the blocker_height from blockers when necessary instead of requiring blockers to update CChainState every time their blocker_height changes?
> 
> That won't work for external or offline blockers. In #19463, my primary goal is to support tools like Armory or Lightning nodes that require parsing the block data, as well as ensuring pruning doesn't invalidate wallet backups (even of offline wallets).

The way I think this would work is: if `CChainState` asks `WalletBlocker` about the lowest prunable block, and the wallet is offline, `WalletBlocker` would just reply with the last one it knows about. I see in your original PR you planned on persisting blockers, so in my mind upon initialization, `WalletBlocker` would just load up the persisted file.

Lmk if that makes sense. In my mind, all else being equal, a "pull" configuration is more likely to be correct since blockers won't have to worry about forgetting to update their `blocker_height`, and avoids the need to constantly notify CChainState.

luke-jr commented:
>The way I think this would work is: if CChainState asks WalletBlocker about the lowest prunable block, and the wallet is offline, WalletBlocker would just reply with the last one it knows about.

No reason it would be limited to wallets, though. It ends up back to BlockManager keeping track of them all - not sure what the issue is with this? It seems perfectly in line with its role?

>In my mind, all else being equal, a "pull" configuration is more likely to be correct since blockers won't have to worry about forgetting to update their blocker_height, and avoids the need to constantly notify CChainState.

For each block, indexes will move their lock one height later - pull or push doesn't matter in this scenario. But other locks might not move right away, so keeping track in the BlockManager avoids unnecessary interaction between them. It also allows BlockManager to shift backward any locks that will be impacted by a reorganization, preventing pruning of those blocks before the clients even know they need an older height.

jonatack commented:
> That is, unless the test is now too slow to run on some machines now. Please test it again on your toaster, @jonatack ;)

Happy to :) ... needs rebase (again).

fjahr commented on src/validation.h:
```diff
@@ -455,6 +458,16 @@ class BlockManager
+     */
+    std::unordered_map<std::string, const CBlockIndex*> m_prune_blockers GUARDED_BY(::cs_main);
+
+    void UpdatePruneBlocker(const std::string& name, const CBlockIndex* block) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
```
I slightly prefer blocker because lock makes me think of mutex locks, which is even present on the same line. So I have kept the term blockers.

luke-jr replied:
Mutex locks and prune locks are fundamentally the same thing...?

fjahr replied:
Yes, to me that doesn't make it less confusing

luke-jr replied:
Using the same word (lock) for the same thing makes sense.

Using the same word (block) for completely different things, is confusing.

fjahr replied:
Renamed to locks.

fjahr commented:
Rebased and addressed @luke-jr comments (now co-author of the main commit introducing prune blockers).

fjahr commented:
> The RPC interface for prune locks (not in this PR presently) could be useful for testing, so it might be a good idea to add it. If you want, you can just cherry-pick [9c9e0c0](https://github.com/bitcoin/bitcoin/commit/9c9e0c0b8c6a5a332f45896261aca5bdb3746447) + [e83ed21](https://github.com/bitcoin/bitcoin/commit/e83ed21172049ef593702c52bb840868728f47d1) + [da86f7b](https://github.com/bitcoin/bitcoin/commit/da86f7bdcfda0f5b27091b3934bc9f3e8de03157) for that.

I have not done this yet because I don't want to expand the scope of this PR too much. But I am happy to open a new PR with these commits as a follow-up when this one here has been merged.

jonatack commented on src/validation.h:
```diff
@@ -35,6 +35,7 @@
 #include <stdint.h>
 #include <string>
 #include <thread>
+#include <unordered_map>
```
b44fea2c it looks like these two includes should be added to `node/blockstorage.{h,cpp}` -- or only the header file if you inline `UpdatePruneBlocker()` in the h file -- instead of the `validation.{h, cpp} files`

fjahr replied:
Done

jonatack commented on src/validation.cpp:
```diff
@@ -2468,6 +2474,18 @@ bool CChainState::DisconnectTip(BlockValidationState& state, DisconnectedBlockTr
+        const int max_height_first = pindexDelete->nHeight - 1;
+        for (auto& prune_blocker : m_blockman.m_prune_blockers) {
+            if (prune_blocker.second.m_height_first <= max_height_first) continue;
+            // NOTE: Don't need to write to db here, since it will get synced when the rest of the chainstate
```
b44fea2c This comment seems to be unfinished or missing context ("when"...what?), and not sure if "when" should be "with"

fjahr replied:
Done

jonatack commented on src/index/base.cpp:
```diff
@@ -376,3 +372,15 @@ IndexSummary BaseIndex::GetSummary() const
+        node::PruneBlockerInfo prune_blocker;
+        prune_blocker.m_height_first = block->nHeight;
+        LOCK(::cs_main);
+        m_chainstate->m_blockman.UpdatePruneBlocker(GetName(), prune_blocker);
```
a879711
```diff
-        LOCK(::cs_main);
-        m_chainstate->m_blockman.UpdatePruneBlocker(GetName(), prune_blocker);
+        WITH_LOCK(::cs_main, m_chainstate->m_blockman.UpdatePruneBlocker(GetName(), prune_blocker));
```

fjahr replied:
Done

jonatack commented on src/index/base.h:
```diff
@@ -103,6 +103,8 @@ class BaseIndex : public CValidationInterface
     /// Get the name of the index for display in logs.
     virtual const char* GetName() const = 0;
 
+    virtual bool AllowPrune() const = 0;
```
a879711 There doesn't seem to be any need for the four declarations of `AllowPrune()` to be protected; they can just be private?

fjahr replied:
Done

jonatack commented on src/init.cpp:
```diff
@@ -417,7 +417,7 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-persistmempool", strprintf("Whether to save the mempool on shutdown and load on restart (default: %u)", DEFAULT_PERSIST_MEMPOOL), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-pid=<file>", strprintf("Specify pid file. Relative paths will be prefixed by a net-specific datadir location. (default: %s)", BITCOIN_PID_FILENAME), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
-    argsman.AddArg("-prune=<n>", strprintf("Reduce storage requirements by enabling pruning (deleting) of old blocks. This allows the pruneblockchain RPC to be called to delete specific blocks, and enables automatic pruning of old blocks if a target size in MiB is provided. This mode is incompatible with -txindex and -coinstatsindex. "
+    argsman.AddArg("-prune=<n>", strprintf("Reduce storage requirements by enabling pruning (deleting) of old blocks. This allows the pruneblockchain RPC to be called to delete specific blocks, and enables automatic pruning of old blocks if a target size in MiB is provided. This mode is incompatible with -txindex. "
```
16f6718
```suggestion
    argsman.AddArg("-prune=<n>", strprintf("Reduce storage requirements by enabling pruning (deleting) of old blocks. This allows the pruneblockchain RPC to be called to delete specific blocks and enables automatic pruning of old blocks if a target size in MiB is provided. This mode is incompatible with -txindex. "
```

fjahr replied:
Done

jonatack reviewed (commented):
Re-reviewing from scratch. Haven't yet re-reviewed the tests.

I still need to run `test/functional/feature_index_prune.py --timeout-factor=0` to avoid timeouts and the test still takes 15-16 minutes to run in performance mode with my cores at max frequency, so no change from https://github.com/bitcoin/bitcoin/pull/21726#discussion_r723429759 and the suggestion to place the test in the extended scripts of the test runner (or at the top of the base scripts, or speed it up). 

Edit: The slowest test is `make sure the nodes start again with the indices and an additional -reindex arg`, taking 8-9 minutes out of 16 for me, or a bit more than half of the total test run time.

fjahr commented:
> I still need to run `test/functional/feature_index_prune.py --timeout-factor=0` to avoid timeouts and the test still takes 15-16 minutes to run in performance mode with my cores at max frequency, so no change from [#21726 (comment)](https://github.com/bitcoin/bitcoin/pull/21726#discussion_r723429759) and the suggestion to place the test in the extended scripts of the test runner (or at the top of the base scripts, or speed it up).

That is actually an improvement relatively speaking because for me the test became about 2.5x slower since the rebase [here](https://github.com/bitcoin/bitcoin/pull/21726#issuecomment-996289321). For me the other extended tests are still a lot slower than this one but it's not too far off anymore so I have added the test to the list now.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -1115,7 +1115,7 @@ static RPCHelpMan pruneblockchain()
                 },
                 RPCResult{
-                    RPCResult::Type::NUM, "", "Height of the last block pruned"},
+                    RPCResult::Type::NUM, "", "The height of the first block that is still stored after pruning"},
```
Done

fjahr commented on src/node/blockstorage.cpp:
```diff
@@ -51,6 +51,15 @@ bool IsBlockPruned(const CBlockIndex* pblockindex)
+const CBlockIndex* GetFirstStoredBlock(const CBlockIndex* start_block) {
+    CHECK_NONFATAL(start_block);
+    const CBlockIndex* last_block = start_block;
+    while (last_block->pprev && (last_block->pprev->nStatus & BLOCK_HAVE_DATA)) {
```
Done

fjahr commented:
Addressed comments by @jonatack and renamed from blockers to locks since I am not super passionate about naming stuff and I hope getting rid of controversies gets this closer to a merge faster.

jonatack commented:
Thanks for updating @fjahr, re-reviewing soon.

jonatack reviewed (commented):
Updates LGTM. I need to go through the functional tests to finish the review. One WIP comment.

mzumsande commented on src/index/base.cpp:
```diff
@@ -372,3 +372,14 @@ IndexSummary BaseIndex::GetSummary() const
+    if (AllowPrune() && block) {
+        node::PruneLockInfo prune_lock;
+        prune_lock.m_height_first = block->nHeight;
+        WITH_LOCK(::cs_main, m_chainstate->m_blockman.UpdatePruneLock(GetName(), prune_lock));
```
This updates the prune lock for coinstatsindex and blockfilterindex regularly, even if we don't have pruning enabled. Is that necessary?

fjahr replied:
I was feeling the same and previously had a commit included here but removed it due to feedback that it wasn't needed: https://github.com/bitcoin/bitcoin/commit/e365e87a72cc49e175b31256a11a6c5b1621d1f2

I guess I could open a follow-up PR but for now I am trying to keep this PR as simple as possible to move it forward.

mzumsande commented on src/validation.cpp:
```diff
@@ -2468,6 +2473,18 @@ bool CChainState::DisconnectTip(BlockValidationState& state, DisconnectedBlockTr
+    {
+        // Prune locks that began at or after the tip should get moved backward so they get a chance to reorg
+        const int max_height_first = pindexDelete->nHeight - 1;
+        for (auto& prune_lock : m_blockman.m_prune_locks) {
```
Couldn't this also be skipped when we are not pruning?

fjahr replied:
(see my other comment, same reasoning)

mzumsande reviewed (commented):
Concept ACK. 
The code looks good to me on first pass, my only question is why bookkeeping for prune locks is done when not pruning (see below).

luke-jr commented:
We don't really have a "never prune" option right now, just a "haven't pruned yet, no plans to do so" which could be changed at any time by the user...

mzumsande commented:
> We don't really have a "never prune" option right now, just a "haven't pruned yet, no plans to do so" which could be changed at any time by the user...

Right, but we do have a `fPruneMode` flag that is only set in Init, so we can't switch to pruning without a restart afaik. And since the spot where the prune locks are read is conditional on `fPruneMode` ([code](https://github.com/bitcoin/bitcoin/pull/21726/commits/e530e5e76a0c61d989dec98d7443deeccf334148#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98L2227)) I was wondering whether it would make sense to do the same with the spots where the locks are written.
The init code must ensure anyway that nothing is pruned at startup before the indexes are initialized and the prune blockers are set.

luke-jr commented:
I guess this PR doesn't persist prune locks at all, but that is planned for afterward, so I'm not sure it makes sense to special-case it here (which would also be a layering violation IMO).

luke-jr commented on src/validation.cpp:
```diff
@@ -2468,6 +2473,18 @@ bool CChainState::DisconnectTip(BlockValidationState& state, DisconnectedBlockTr
+        const int max_height_first = pindexDelete->nHeight - 1;
+        for (auto& prune_lock : m_blockman.m_prune_locks) {
+            if (prune_lock.second.m_height_first <= max_height_first) continue;
+            // NOTE: Don't need to write to db here, since it will get synced with the rest of the chainstate
```
Comment is premature since this PR never writes prune locks :)

fjahr replied:
Removed

luke-jr commented on src/validation.cpp:
```diff
@@ -2225,12 +2224,18 @@ bool CChainState::FlushStateToDisk(
+            // make sure we don't prune above any of the prune locks bestblocks
             // pruning is height-based
             int last_prune = m_chain.Height(); // last height we can prune
-            ForEachBlockFilterIndex([&](BlockFilterIndex& index) {
```
This is deleted one commit too early (part of "blockstorage: Add prune locks to BlockManager" rather than "Index: Use prune locks for blockfilterindex")

fjahr replied:
fixed

luke-jr commented on src/init.cpp:
```diff
@@ -867,12 +867,10 @@ bool AppInitParameterInteraction(const ArgsManager& args)
             return InitError(_("Prune mode is incompatible with -txindex."));
-        if (args.GetBoolArg("-coinstatsindex", DEFAULT_COINSTATSINDEX))
-            return InitError(_("Prune mode is incompatible with -coinstatsindex."));
     }
```
IMO this should be a separate PR so we don't have to review the whole coinstatsindex logic for prune-safety in this PR.

fjahr replied:
Hmm, this part was, to me, the main motivation for this PR. Given the review that has already gone into it up until now with this included I think it's ok to leave it in. But will reconsider if others agree with you.

fjahr commented on src/validation.h:
```diff
@@ -100,6 +100,8 @@ static const unsigned int DEFAULT_CHECKLEVEL = 3;
 // Setting the target to >= 550 MiB will make it likely we can respect the target.
 static const uint64_t MIN_DISK_SPACE_FOR_BLOCK_FILES = 550 * 1024 * 1024;
+/** The number of blocks to keep below the deepest prune lock. */
+static constexpr int PRUNE_LOCK_BUFFER{10};
```
The number is kind of arbitrary: A bit higher than what we expect to see on mainnet, like at this level we might actually have bigger problems than the indices on pruned nodes. On the other hand low enough to not have too much impact on the pruning mechanism. I have added a bit more concise version of this to the comment.

Good catch on the commit message, fixed that as well.

jonatack replied:
`PRUNE_LOCK_BUFFER` is only used in validation.cpp, so it could perhaps be moved there.

fjahr commented on src/validation.cpp:
```diff
@@ -2225,12 +2224,18 @@ bool CChainState::FlushStateToDisk(
+                if (prune_lock.second.m_height_first == std::numeric_limits<int>::max()) continue;
+                const int lock_height{prune_lock.second.m_height_first - PRUNE_LOCK_BUFFER};
+                last_prune = std::max(1, std::min(last_prune, lock_height));
+                if (last_prune == lock_height) {
```
Yeah, makes sense, changed the logging part.

fjahr commented on src/node/blockstorage.h:
```diff
@@ -156,6 +161,16 @@ class BlockManager
     CBlockIndex* GetLastCheckpoint(const CCheckpointData& data) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
+    /**
+     * Map from external index name to most recent block the index can tolerate being pruned.
```
done

fjahr commented:
I hope I addressed all the comments, thanks for reviewing!

ryanofsky commented on src/validation.cpp:
```diff
@@ -2225,13 +2224,28 @@ bool CChainState::FlushStateToDisk(
 
+            for (const auto& prune_lock : m_blockman.m_prune_locks) {
+                if (prune_lock.second.m_height_first == std::numeric_limits<int>::max()) continue;
+                const int lock_height{prune_lock.second.m_height_first - PRUNE_LOCK_BUFFER};
```
In commit "blockstorage: Add prune locks to BlockManager" (d249636878b39aa34222d335cd85f0e2561bd50a)

I think there is a minor off-by-one error here and this should be `lock_height = height_first - PRUNE_LOCK_BUFFER - 1`.

If `height_first` is 100 ("oldest block that must not be pruned"), and `PRUNE_LOCK_BUFFER` is 1 ("number of blocks to keep below the deepest prune lock"), then `last_prune` ("last height we can prune") should be `98` not `99`. Pruning up to 99 when 100 is the first block that can't be pruned would be leaving a buffer of size 0 not size 1.

fjahr replied:
Yeah, makes sense, I also added a comment here to make it clearer to future readers what the additional `- 1` is there for.

ryanofsky commented on src/validation.cpp:
```diff
@@ -2468,6 +2482,18 @@ bool CChainState::DisconnectTip(BlockValidationState& state, DisconnectedBlockTr
     LogPrint(BCLog::BENCH, "- Disconnect block: %.2fms\n", (GetTimeMicros() - nStart) * MILLI);
+
+    {
+        // Prune locks that began at or after the tip should get moved backward so they get a chance to reorg
```
In commit "blockstorage: Add prune locks to BlockManager" (d249636878b39aa34222d335cd85f0e2561bd50a)

Note just for understanding this: Interesting to see this logic added, since it wasn't in original version of the PR. It looks like the result of discussion with Luke in https://github.com/bitcoin/bitcoin/pull/21726#discussion_r785479282 and is it checked by a python test looking for the "lock moved back to" log message below.

IMO, this change adds a little bit of internal complexity but improves the external design and reliability of prune locks. If you think about indexes running asynchronously behind the main chain, it's a nice thing if prune locks are updated automatically when there's a reorg, and indexes are free to catch up to block connected and disconnected notifications at any time later. They don't need to suddenly hurry and update their locks because this code does it for them.

ryanofsky reviewed (approved):
Code review ACK eccb1b4b0b3bf26850e73af1c33ad65980730918. Main change since last review is PR has been Lukified with two nice changes: introducing the PruneLockInfo struct which should make locks more extensible, and automatically updating locks on reorgs so indexes (and wallets in the future) can be safe after reorgs if they are running more asynchronously. There were also a lot of rebase fixes and other smaller changes.

fjahr commented on src/node/blockstorage.h:
```diff
@@ -156,6 +161,16 @@ class BlockManager
+    /**
+     * Map from external index name to oldest block that must not be pruned.
+     *
+     * @note Internally, only blocks at height (block->nHeight - PRUNE_LOCK_BUFFER) and
```
done

fjahr commented on src/node/blockstorage.h:
```diff
@@ -55,6 +56,10 @@ struct CBlockIndexWorkComparator {
 };
 
+struct PruneLockInfo {
+    int m_height_first{std::numeric_limits<int>::max()};
```
Makes sense, done (both)

fjahr commented:
Addressed comments from @ryanofsky for some nice improvements , thanks!

fjahr commented:
Rebased

luke-jr commented on src/rpc/blockchain.cpp:
```diff
@@ -744,7 +744,7 @@ static RPCHelpMan pruneblockchain()
                 },
                 RPCResult{
-                    RPCResult::Type::NUM, "", "Height of the last block pruned"},
+                    RPCResult::Type::NUM, "", "The height of the first (lowest) block that is still stored, after pruning"},
```
This is actually a bug. From 0.14 (2017 Mar) until before 0.19 (2019 Nov), the height of the last block pruned was returned, subject to a bug if there were blocks left unpruned due to sharing files with later blocks.

In #15991, this was "fixed" to the current implementation, introducing the new bug you're documenting here.

Since the user provides the parameter as a block to include in pruning, I have a slight preference for fixing the behaviour to match the documentation. See #24629

fjahr replied:
Ok, makes sense. Dropped the commit and will take a look at #24629.

ryanofsky reviewed (approved):
Code review ACK ce081ba2b1a2c4811df9a59e752d00bfcfce35fe. Only changes since last review were rebasing and making suggested cleanups, including fixing off-by-one buffer value.

I do think it would be good to drop commit "RPC: Fix pruneblockchain return documentation" (f2d397c68e5042dee5827402d27741f1a89ff583), though, for reasons Luke suggested

mzumsande reviewed (commented):
Code review ACK ce081ba2b1a2c4811df9a59e752d00bfcfce35fe

This looks good to me, just two small comments below - happy to reACK after rebase / possibly removing the pruneblockchain doc change commit.

fjahr commented on src/validation.cpp:
```diff
@@ -2277,13 +2277,29 @@ bool CChainState::FlushStateToDisk(
             });
 
+            for (const auto& prune_lock : m_blockman.m_prune_locks) {
+                if (prune_lock.second.height_first == std::numeric_limits<int>::max()) continue;
```
Yeah, at the current state of the code I don't see how this should happen. But I still would like to keep something there as belts & suspenders. But since this would be a developer error and not a usual case, I changed this line to an `assert` because it seems more appropriate here.

fjahr commented on src/node/blockstorage.cpp:
```diff
@@ -397,6 +397,16 @@ bool IsBlockPruned(const CBlockIndex* pblockindex)
 
+const CBlockIndex* GetFirstStoredBlock(const CBlockIndex* start_block) {
+    AssertLockHeld(::cs_main);
+    CHECK_NONFATAL(start_block);
```
I agree, changed.

fjahr commented:
Rebased and addressed latest comments

ryanofsky commented on src/node/blockstorage.cpp:
```diff
@@ -397,6 +397,15 @@ bool IsBlockPruned(const CBlockIndex* pblockindex)
+const CBlockIndex* GetFirstStoredBlock(const CBlockIndex* start_block) {
+    AssertLockHeld(::cs_main);
+    const CBlockIndex* last_block = start_block;
+    while (last_block->pprev && (last_block->pprev->nStatus & BLOCK_HAVE_DATA)) {
```
In commit "refactor: Introduce GetFirstStoredBlock helper function" (8323cce2aa76123bc385c7927701d42282b0bd26)

Previous version of this function in 19610999595ec557fc602c8cefb930c2dfc1ad67 had a `CHECK_NONFATAL(start_block)` line, which I think was good because it prevented UB if a null start block was pass, which seems like an easy mistake to make. IMO it would be nice to restore this, or add `assert(start_block)` since callers are now using CHECK_NONFATAL.

fjahr replied:
I think the `assert` makes sense here now, added.

ryanofsky commented on src/validation.cpp:
```diff
@@ -2270,13 +2270,30 @@ bool CChainState::FlushStateToDisk(
 
+            for (const auto& prune_lock : m_blockman.m_prune_locks) {
+                assert(prune_lock.second.height_first < std::numeric_limits<int>::max());
+                // if (prune_lock.second.height_first == std::numeric_limits<int>::max()) continue;
```
In commit "blockstorage: Add prune locks to BlockManager" (a473db7e6a97289f981c3ff4b560e07a680efed5)

Could drop the commented line

fjahr replied:
🤦‍♂️  done

ryanofsky reviewed (approved):
Code review ACK 735e16c8e2e6e7023c58570cfc55b8f031136ea9. Main change since last review is dropping commit f2d397c68e5042dee5827402d27741f1a89ff583 "RPC: Fix pruneblockchain return documentation". Also rebased and tweaked "limited pruning to height" log message and `GetFirstStoredBlock` null block asserts

ryanofsky reviewed (approved):
Code review ACK 2faa5b51799e690e20e25e660d7b18a18ae4dc62. Just suggested comment and assert tweaks since last review

mzumsande reviewed (approved):
Code review ACK 2faa5b51799e690e20e25e660d7b18a18ae4dc62

mzumsande commented on src/index/base.cpp:
```diff
@@ -75,11 +75,7 @@ bool BaseIndex::Init()
-                block = block->pprev;
-            }
-            prune_violation = block != active_chain.Genesis();
+            prune_violation = node::GetFirstStoredBlock(active_chain.Tip()) != active_chain.Genesis();
```
Just noticing that the reason this cannot not trigger the added assert in `GetFirstStoredBlock()` is that if `active_chain.Tip()` is nullptr (which happens on first startup), then `m_synced` will be determined to be true, and we don't get to this point (#22485 documents this). I think that this is a bit fragile, but doesn't need to be addressed in this PR.

jonatack commented on src/node/blockstorage.h:
```diff
@@ -171,6 +176,16 @@ class BlockManager
+     * @note Internally, only blocks at height (height_first - PRUNE_LOCK_BUFFER) and
+     * below will be pruned, but callers should avoid assuming any particular buffer size.
+     */
+    std::unordered_map<std::string, PruneLockInfo> m_prune_locks GUARDED_BY(::cs_main);
```
527ef44 It looks like `m_prune_locks` can be private instead of public.

fjahr replied:
Right, done.

jonatack reviewed (commented):
Re-reviewing. Each commit builds cleanly. Needs rebase. 

Edit: utACK modulo comments and didn't yet review the tests (will do after rebase). 

<details><summary>a couple other minor suggestions so far</summary><p>

```diff
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -2271,7 +2277,7 @@ bool CChainState::FlushStateToDisk(
         if (fPruneMode && (m_blockman.m_check_for_pruning || nManualPruneHeight > 0) && !fReindex) {
             // make sure we don't prune above any of the prune locks bestblocks
             // pruning is height-based
-            int last_prune = m_chain.Height(); // last height we can prune
+            int last_prune{m_chain.Height()}; // last height we can prune
             std::optional<std::string> limiting_lock; // prune lock that actually was the limiting factor, only used for logging
 
             for (const auto& prune_lock : m_blockman.m_prune_locks) {
@@ -2526,8 +2532,8 @@ bool CChainState::DisconnectTip(BlockValidationState& state, DisconnectedBlockTr
     LogPrint(BCLog::BENCH, "- Disconnect block: %.2fms\n", (GetTimeMicros() - nStart) * MILLI);
 
     {
-        // Prune locks that began at or after the tip should get moved backward so they get a chance to reorg
-        const int max_height_first = pindexDelete->nHeight - 1;
+        // Prune locks that began at or after the tip should be moved backward so they have a chance to reorg
+        const int max_height_first{pindexDelete->nHeight - 1};
         for (auto& prune_lock : m_blockman.m_prune_locks) {
             if (prune_lock.second.height_first <= max_height_first) continue;
```
</p></details>

jonatack commented on src/index/base.h:
```diff
@@ -125,6 +128,8 @@ class BaseIndex : public CValidationInterface
     /// Get a summary of the index and its state.
     IndexSummary GetSummary() const;
+
+    void SetBestBlockIndex(const CBlockIndex* block);
```
3b8b898 it looks like `SetBestBlockIndex` can be private

fjahr replied:
Right, done.

fjahr commented:
Rebased and addressed comments from @jonatack, thanks for reviewing!

Also @mzumsande is so kind to host a [PR Review Club on this PR](https://bitcoincore.reviews/21726) this week, in case anyone is interested in joining :)

ryanofsky reviewed (approved):
Code review ACK f37f2020eb3bd782206de7532b428b29e963718f. Changes since last review: cleanup to comments and initialization, test updates

luke-jr commented on src/validation.cpp:
```diff
@@ -2339,12 +2344,24 @@ bool CChainState::FlushStateToDisk(
+            std::optional<std::string> limiting_lock; // prune lock that actually was the limiting factor, only used for logging
+
+            for (const auto& prune_lock : m_blockman.m_prune_locks) {
+                assert(prune_lock.second.height_first < std::numeric_limits<int>::max());
```
I don't think this should assert. Seems like begging for a DoS vector.

fjahr replied:
It seems unlikely to me but for the sake of safety I changed it back to just `continue`.

fjahr commented on src/node/blockstorage.h:
```diff
@@ -120,6 +125,14 @@ class BlockManager
+    /**
+     * Map from external index name to oldest block that must not be pruned.
+     *
+     * @note Internally, only blocks at height (height_first - PRUNE_LOCK_BUFFER) and
```
Yepp, fixed.

luke-jr commented on src/node/blockstorage.h:
```diff
@@ -175,12 +188,20 @@ class BlockManager
     }
 };
 
+//! Check whether the block associated with this index entry is pruned or not.
+bool IsBlockPruned(const CBlockIndex* pblockindex) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
```
Bad rebase?

fjahr replied:
Yepp

fjahr commented:
Another rebase, removed a line that slipped in during the last rebase and added a small comment.

ryanofsky reviewed (approved):
Code review ACK 71c3f0356c01521a95c64fba1e7375aea6286bb0. Changes since last review: just tweaking comments and asserts, and rebasing

w0xlt reviewed (approved):
tACK https://github.com/bitcoin/bitcoin/pull/21726/commits/71c3f0356c01521a95c64fba1e7375aea6286bb0 on signet.

MarcoFalke commented on src/node/blockstorage.cpp:
```diff
@@ -397,6 +404,16 @@ bool BlockManager::IsBlockPruned(const CBlockIndex* pblockindex)
 
+const CBlockIndex* GetFirstStoredBlock(const CBlockIndex* start_block) {
+    AssertLockHeld(::cs_main);
+    assert(start_block);
```
Why is this function accepting a nullptr, but then aborting on it. Wouldn't it be better to not accept a nullptr (aka reference)?

jonatack replied:
> Why is this function accepting a nullptr, but then aborting on it. Wouldn't it be better to not accept a nullptr (aka reference)?

Proposed in #24150.

MarcoFalke commented on src/node/blockstorage.cpp:
```diff
@@ -397,6 +404,16 @@ bool BlockManager::IsBlockPruned(const CBlockIndex* pblockindex)
+    while (last_block->pprev && (last_block->pprev->nStatus & BLOCK_HAVE_DATA)) {
+        last_block = last_block->pprev;
+    }
+    return last_block;
```
Would be nice if there was a way to annotate the returned value is never nullptr

MarcoFalke replied:
I guess this can be achieved by deleting the move/copy constructor of `CBlockIndex` and then returning a reference instead of pointer.

jonatack replied:
> I guess this can be achieved by deleting the move/copy constructor of `CBlockIndex` and then returning a reference instead of pointer.

interesting

MarcoFalke commented on src/node/blockstorage.h:
```diff
@@ -175,12 +188,18 @@ class BlockManager
 };
 
+//! Find the first block that is not pruned
+const CBlockIndex* GetFirstStoredBlock(const CBlockIndex* start_block) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
```
Why is this a global function, when it could be a member of BlockManager? I think that'd make more sense, if we longer term want to remove the cs_main lock on storage stuff and give BlockManager its own mutex to take.

jonatack replied:
> Why is this a global function, when it could be a member of BlockManager? I think that'd make more sense, if we longer term want to remove the cs_main lock on storage stuff and give BlockManager its own mutex to take.

Done in #24150.

MarcoFalke commented on src/node/blockstorage.cpp:
```diff
@@ -397,6 +404,16 @@ bool BlockManager::IsBlockPruned(const CBlockIndex* pblockindex)
     return (m_have_pruned && !(pblockindex->nStatus & BLOCK_HAVE_DATA) && pblockindex->nTx > 0);
 }
 
+const CBlockIndex* GetFirstStoredBlock(const CBlockIndex* start_block) {
```
nit: clang-format

jonatack replied:
Picked up these clang-formatting nits inside 8507260fd32914fa63 in #24150.

MarcoFalke commented on src/node/blockstorage.cpp:
```diff
@@ -230,6 +232,11 @@ void BlockManager::FindFilesToPrune(std::set<int>& setFilesToPrune, uint64_t nPr
            nLastBlockWeCanPrune, count);
 }
 
+void BlockManager::UpdatePruneLock(const std::string& name, const PruneLockInfo& lock_info) {
```
same nit

MarcoFalke commented on src/index/base.cpp:
```diff
@@ -381,3 +377,14 @@ IndexSummary BaseIndex::GetSummary() const
     return summary;
 }
+
+void BaseIndex::SetBestBlockIndex(const CBlockIndex* block) {
```
same nit

MarcoFalke commented on src/index/base.cpp:
```diff
@@ -138,7 +134,7 @@ void BaseIndex::ThreadSync()
         int64_t last_locator_write_time = 0;
         while (true) {
             if (m_interrupt) {
-                m_best_block_index = pindex;
```
Haven't tried this, but what happens if `m_best_block_index` is private?

jonatack replied:
Pulling some of these into #24150 (as it needs to be rebased anyway now).

fjahr replied:
I'm not sure I can follow: `m_best_block_index` is already private?

maflcko reviewed (commented):
some nits and questions
